File: vscode/src/queryExecutor.ts
import * as vscode from 'vscode';
import { Client } from 'pg';
import { QueryParameter } from './openRouterClient';

export interface QueryResult {
    columns: string[];
    rows: any[][];
    rowCount: number;
    durationMs: number;
    affectedRows: number;
    isSelect: boolean;
    error?: string;
}

export class QueryExecutor {
    private getConnectionString(): string | undefined {
        return vscode.workspace.getConfiguration('psqlQueryTester').get('connectionString');
    }

    private getTimeout(): number {
        return vscode.workspace.getConfiguration('psqlQueryTester').get('queryTimeout') || 30;
    }

    private getMaxRows(): number {
        return vscode.workspace.getConfiguration('psqlQueryTester').get('maxResultRows') || 1000;
    }

    async execute(sql: string, parameters: QueryParameter[], parameterValues: Record<string, any>): Promise<QueryResult> {
        const startTime = Date.now();

        if (!sql.trim()) {
            return {
                columns: [],
                rows: [],
                rowCount: 0,
                durationMs: 0,
                affectedRows: 0,
                isSelect: true,
                error: 'No SQL query to execute.'
            };
        }

        const connectionString = this.getConnectionString();
        if (!connectionString) {
            return {
                columns: [],
                rows: [],
                rowCount: 0,
                durationMs: 0,
                affectedRows: 0,
                isSelect: true,
                error: 'Database not connected. Please configure connection in Settings > PSQL Query Tester.'
            };
        }

        const client = new Client({ connectionString });

        try {
            await client.connect();

            // Set statement timeout
            const timeoutSec = this.getTimeout();
            await client.query('SET statement_timeout = \'' + timeoutSec + 's\'');

            // Build parameter values array in correct order
            const values = parameters
                .sort((a, b) => a.position - b.position)
                .map(p => this.convertValue(parameterValues[p.name], p.type));

            const isSelect = sql.trim().toUpperCase().startsWith('SELECT');

            const result = await client.query(sql, values);
            const durationMs = Date.now() - startTime;

            if (isSelect) {
                const columns = result.fields.map(f => f.name);
                const rows = result.rows.slice(0, this.getMaxRows()).map(row => 
                    columns.map(col => this.formatValue(row[col]))
                );

                return {
                    columns,
                    rows,
                    rowCount: result.rowCount || 0,
                    durationMs,
                    affectedRows: 0,
                    isSelect: true
                };
            } else {
                return {
                    columns: ['Affected Rows'],
                    rows: [[result.rowCount || 0]],
                    rowCount: 1,
                    durationMs,
                    affectedRows: result.rowCount || 0,
                    isSelect: false
                };
            }
        } catch (e: any) {
            return {
                columns: [],
                rows: [],
                rowCount: 0,
                durationMs: Date.now() - startTime,
                affectedRows: 0,
                isSelect: true,
                error: 'Query execution failed: ' + e.message
            };
        } finally {
            await client.end();
        }
    }

    async executeSimple(sql: string): Promise<QueryResult> {
        return this.execute(sql, [], {});
    }

    private convertValue(value: any, type: string): any {
        if (value === null || value === undefined || value === '') {
            return null;
        }

        switch (type.toLowerCase()) {
            case 'integer':
            case 'int':
            case 'int4':
            case 'bigint':
            case 'int8':
                return parseInt(value, 10);
            case 'boolean':
            case 'bool':
                return value === true || value === 'true' || value === '1';
            case 'numeric':
            case 'decimal':
            case 'float':
            case 'double':
            case 'real':
                return parseFloat(value);
            default:
                return value;
        }
    }

    private formatValue(value: any): string {
        if (value === null || value === undefined) {
            return 'NULL';
        }
        if (value instanceof Date) {
            return value.toISOString();
        }
        if (typeof value === 'object') {
            return JSON.stringify(value);
        }
        return String(value);
    }
}

export async function getSchema(): Promise<string | undefined> {
    const connectionString = vscode.workspace.getConfiguration('psqlQueryTester').get<string>('connectionString');
    if (!connectionString) {
        return undefined;
    }

    const client = new Client({ connectionString });

    try {
        await client.connect();

        const result = await client.query(`
            SELECT 
                t.table_name,
                c.column_name,
                c.data_type,
                c.is_nullable
            FROM information_schema.tables t
            JOIN information_schema.columns c 
                ON t.table_name = c.table_name 
                AND t.table_schema = c.table_schema
            WHERE t.table_schema = 'public'
                AND t.table_type = 'BASE TABLE'
            ORDER BY t.table_name, c.ordinal_position
        `);

        const tables: Record<string, string[]> = {};
        for (const row of result.rows) {
            const tableName = row.table_name;
            if (!tables[tableName]) {
                tables[tableName] = [];
            }
            const nullable = row.is_nullable === 'NO' ? ', NOT NULL' : '';
            tables[tableName].push(row.column_name + ' (' + row.data_type + nullable + ')');
        }

        let schema = 'DATABASE SCHEMA:\n\n';
        for (const [table, columns] of Object.entries(tables)) {
            schema += 'TABLE: ' + table + '\n';
            schema += columns.map(c => '  - ' + c).join('\n');
            schema += '\n\n';
        }

        return schema;
    } catch (e) {
        console.error('Failed to get schema:', e);
        return undefined;
    } finally {
        await client.end();
    }
}

File: src/main/kotlin/com/psqlquerytester/db/QueryExecutor.kt
package com.psqlquerytester.db

import com.psqlquerytester.api.QueryParameter
import com.psqlquerytester.settings.PluginSettings
import java.sql.ResultSet
import java.sql.Types

data class QueryResult(
    val columns: List<String>,
    val rows: List<List<Any?>>,
    val rowCount: Int,
    val durationMs: Long,
    val affectedRows: Int = 0,
    val isSelect: Boolean = true,
    val error: String? = null
)

object QueryExecutor {

    /**
     * Convert PostgreSQL-style $1, $2 parameters to JDBC-style ? parameters
     */
    private fun convertToJdbcParams(sql: String): String {
        // Replace $1, $2, $3, etc. with ?
        return sql.replace(Regex("""\$\d+"""), "?")
    }

    /**
     * Execute a simple query without parameters (used for AI-generated parameter value queries)
     */
    fun executeSimple(sql: String): QueryResult {
        val startTime = System.currentTimeMillis()

        if (sql.isBlank()) {
            return QueryResult(
                columns = emptyList(),
                rows = emptyList(),
                rowCount = 0,
                durationMs = 0,
                error = "No SQL query to execute."
            )
        }

        val connection = ConnectionManager.getConnection()
            ?: return QueryResult(
                columns = emptyList(),
                rows = emptyList(),
                rowCount = 0,
                durationMs = 0,
                error = "Database not connected."
            )

        try {
            val settings = PluginSettings.getInstance()
            connection.use { conn ->
                conn.createStatement().use { stmt ->
                    stmt.queryTimeout = settings.queryTimeoutSeconds

                    val isSelect = sql.trim().uppercase().startsWith("SELECT")

                    if (isSelect) {
                        stmt.executeQuery(sql).use { rs ->
                            val result = processResultSet(rs, 1) // Only need first row
                            val duration = System.currentTimeMillis() - startTime
                            return result.copy(durationMs = duration)
                        }
                    } else {
                        return QueryResult(
                            columns = emptyList(),
                            rows = emptyList(),
                            rowCount = 0,
                            durationMs = System.currentTimeMillis() - startTime,
                            error = "AI assist queries must be SELECT statements."
                        )
                    }
                }
            }
        } catch (e: Exception) {
            val duration = System.currentTimeMillis() - startTime
            return QueryResult(
                columns = emptyList(),
                rows = emptyList(),
                rowCount = 0,
                durationMs = duration,
                error = "Query failed: ${e.message}"
            )
        }
    }

    fun execute(
        sql: String,
        parameters: List<QueryParameter>,
        parameterValues: Map<String, Any?>
    ): QueryResult {
        val startTime = System.currentTimeMillis()

        if (sql.isBlank()) {
            return QueryResult(
                columns = emptyList(),
                rows = emptyList(),
                rowCount = 0,
                durationMs = 0,
                error = "No SQL query to execute. The query appears to be empty."
            )
        }

        val connection = ConnectionManager.getConnection()
            ?: return QueryResult(
                columns = emptyList(),
                rows = emptyList(),
                rowCount = 0,
                durationMs = 0,
                error = "Database not connected. Please configure connection in Settings > PSQL Query Tester."
            )

        // Convert PostgreSQL-style $1, $2 parameters to JDBC-style ? parameters
        val jdbcSql = convertToJdbcParams(sql)

        try {
            val settings = PluginSettings.getInstance()
            connection.use { conn ->
                conn.prepareStatement(jdbcSql).use { stmt ->
                    // Set query timeout
                    stmt.queryTimeout = settings.queryTimeoutSeconds

                    // Bind parameters
                    parameters.forEach { param ->
                        val value = parameterValues[param.name]
                        setParameter(stmt, param.position, param.type, value)
                    }

                    // Determine if this is a SELECT query
                    val isSelect = sql.trim().uppercase().startsWith("SELECT")

                    if (isSelect) {
                        stmt.executeQuery().use { rs ->
                            val result = processResultSet(rs, settings.maxResultRows)
                            val duration = System.currentTimeMillis() - startTime
                            return result.copy(durationMs = duration)
                        }
                    } else {
                        val affectedRows = stmt.executeUpdate()
                        val duration = System.currentTimeMillis() - startTime
                        return QueryResult(
                            columns = listOf("Affected Rows"),
                            rows = listOf(listOf(affectedRows)),
                            rowCount = 1,
                            durationMs = duration,
                            affectedRows = affectedRows,
                            isSelect = false
                        )
                    }
                }
            }
        } catch (e: Exception) {
            val duration = System.currentTimeMillis() - startTime
            return QueryResult(
                columns = emptyList(),
                rows = emptyList(),
                rowCount = 0,
                durationMs = duration,
                error = "Query execution failed: ${e.message}"
            )
        }
    }

    private fun setParameter(
        stmt: java.sql.PreparedStatement,
        position: Int,
        type: String,
        value: Any?
    ) {
        if (value == null || (value is String && value.isBlank())) {
            stmt.setNull(position, getSqlType(type))
            return
        }

        when (type.lowercase()) {
            "integer", "int", "int4", "bigint", "int8" -> {
                val intVal = when (value) {
                    is Number -> value.toLong()
                    is String -> value.toLongOrNull() ?: 0L
                    else -> 0L
                }
                stmt.setLong(position, intVal)
            }
            "boolean", "bool" -> {
                val boolVal = when (value) {
                    is Boolean -> value
                    is String -> value.lowercase() in listOf("true", "1", "yes", "t")
                    else -> false
                }
                stmt.setBoolean(position, boolVal)
            }
            "numeric", "decimal", "float", "double", "real", "float4", "float8" -> {
                val doubleVal = when (value) {
                    is Number -> value.toDouble()
                    is String -> value.toDoubleOrNull() ?: 0.0
                    else -> 0.0
                }
                stmt.setDouble(position, doubleVal)
            }
            "timestamp", "timestamptz", "timestamp with time zone" -> {
                val tsVal = when (value) {
                    is java.sql.Timestamp -> value
                    is String -> java.sql.Timestamp.valueOf(value.replace("T", " ").replace("Z", ""))
                    else -> java.sql.Timestamp(System.currentTimeMillis())
                }
                stmt.setTimestamp(position, tsVal)
            }
            "date" -> {
                val dateVal = when (value) {
                    is java.sql.Date -> value
                    is String -> java.sql.Date.valueOf(value)
                    else -> java.sql.Date(System.currentTimeMillis())
                }
                stmt.setDate(position, dateVal)
            }
            "uuid" -> {
                val uuidVal = when (value) {
                    is java.util.UUID -> value
                    is String -> java.util.UUID.fromString(value)
                    else -> java.util.UUID.randomUUID()
                }
                stmt.setObject(position, uuidVal)
            }
            "jsonb", "json" -> {
                stmt.setObject(position, value.toString(), Types.OTHER)
            }
            else -> {
                // Default to string
                stmt.setString(position, value.toString())
            }
        }
    }

    private fun getSqlType(type: String): Int {
        return when (type.lowercase()) {
            "integer", "int", "int4" -> Types.INTEGER
            "bigint", "int8" -> Types.BIGINT
            "boolean", "bool" -> Types.BOOLEAN
            "numeric", "decimal" -> Types.NUMERIC
            "float", "double", "real", "float4", "float8" -> Types.DOUBLE
            "timestamp", "timestamptz", "timestamp with time zone" -> Types.TIMESTAMP
            "date" -> Types.DATE
            "uuid" -> Types.OTHER
            "jsonb", "json" -> Types.OTHER
            else -> Types.VARCHAR
        }
    }

    private fun processResultSet(rs: ResultSet, maxRows: Int): QueryResult {
        val metaData = rs.metaData
        val columnCount = metaData.columnCount

        if (columnCount == 0) {
            return QueryResult(
                columns = listOf("Result"),
                rows = listOf(listOf("Query executed successfully (no columns returned)")),
                rowCount = 1,
                durationMs = 0,
                isSelect = true
            )
        }

        val columns = (1..columnCount).map { metaData.getColumnName(it) }
        val rows = mutableListOf<List<Any?>>()

        var count = 0
        while (rs.next() && count < maxRows) {
            val row = (1..columnCount).map { idx ->
                val value = rs.getObject(idx)
                // Convert some types to more readable formats
                when (value) {
                    is java.sql.Timestamp -> value.toString()
                    is java.sql.Date -> value.toString()
                    is java.sql.Time -> value.toString()
                    is ByteArray -> "[binary data: ${value.size} bytes]"
                    is org.postgresql.util.PGobject -> value.value
                    else -> value
                }
            }
            rows.add(row)
            count++
        }

        if (rows.isEmpty()) {
            return QueryResult(
                columns = columns,
                rows = listOf(listOf("(No rows returned)")),
                rowCount = 0,
                durationMs = 0,
                isSelect = true
            )
        }

        return QueryResult(
            columns = columns,
            rows = rows,
            rowCount = rows.size,
            durationMs = 0,
            isSelect = true
        )
    }
}
